### PR TITLE
Add HDF5 MPI parallel dependency

### PR DESCRIPTION
**Pull request by jorblancoa**
- Added hdf5 mpi parallel and cursesw dependencies
- Modified packages.yaml to point to the new hdf5 library
- Copied BlueConfig file to circuitBuilding directory
- Update Readme to use your username on the ldap_username parameter

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -152,7 +153,8 @@ RUN mkdir $HOME/sim/build \
         && cd circuitBuilding_1000neurons \
         && sed -i "s/CircuitTarget.*/CircuitTarget mini50/g" BlueConfig \
         && sed -i "s/RunMode.*/RunMode RR/g" BlueConfig \
-        && sed -i "s/^Report /#Report /g" BlueConfig
+        && sed -i "s/^Report /#Report /g" BlueConfig \
+	&& cp BlueConfig ../../circuitBuilding_1000neurons
```
Comment (pramodk) [inline Dockerfile]: why we need to copy BlueConfig back? If I am not maitaken we run simulation in the current directory.

**Review by pramodk — APPROVED** [review]
Looks good...could you also new neurodamus modules as well? For example, there are now neurodamus-neocortex, neurodamus-hippocampus and neurodamus-thalamus.

**Comment by pramodk** [comment]
@antonelepfl : could you test this branch?

File: Dockerfile (hunk 1 above)
Comment (jorblancoa) [reply Dockerfile]: Yes, you are totally right. I misread the readme and thought you needed to run the simulation on sim/circuitBuilding, but you do it on sim/build/circuitBuilding that is where the BlueConfig is located...

**Comment by antonelepfl** [comment]
The build process worked correctly but when I try to do 
```
docker run -i -t cellular:latest /bin/bash
cd sim/build/circuitBuilding_1000neurons/
module load neurodamus/master
mpiexec -n 6 --allow-run-as-root special $HOC_LIBRARY_PATH/init.hoc -mpi
```
I get 
```
[6dabb297401a:00077] *** Process received signal ***
[6dabb297401a:00077] Signal: Floating point exception (8)
[6dabb297401a:00077] Signal code: Integer divide-by-zero (1)
[6dabb297401a:00077] Failing at address: 0x7f47765e40a0
[6dabb297401a:00077] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f4776355890]
[6dabb297401a:00077] [ 1] /usr/lib/x86_64-linux-gnu/libopen-pal.so.20(+0x820a0)[0x7f47765e40a0]
[6dabb297401a:00077] [ 2] /usr/lib/x86_64-linux-gnu/libopen-pal.so.20(opal_progress_set_event_poll_rate+0x17)[0x7f4776588b47]
[6dabb297401a:00077] [ 3] /usr/lib/x86_64-linux-gnu/libopen-pal.so.20(opal_progress_init+0x19)[0x7f4776588b89]
[6dabb297401a:00077] [ 4] /usr/lib/x86_64-linux-gnu/libopen-pal.so.20(opal_init+0x151)[0x7f4776589c31]
[6dabb297401a:00077] [ 5] mpiexec(+0x639a)[0x55b38aa5f39a]
[6dabb297401a:00077] [ 6] mpiexec(+0x5aea)[0x55b38aa5eaea]
[6dabb297401a:00077] [ 7] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f4775f73b97]
[6dabb297401a:00077] [ 8] mpiexec(+0x59ea)[0x55b38aa5e9ea]
[6dabb297401a:00077] *** End of error message ***
Floating point exception
```
BTW the only modules that I see are **neurodamus/master** **neuron/2018-10** so I guess I need to use neurodamus/master? before I had to use neurodamus/hippocampus

**Comment by jorblancoa** [comment]
I have pushed new changes with the new neurodamus modules.
However for the circuitBuilding_1000neurons only neurodamus/master and neurodamus-neocortex should work.

@antonelepfl : could you build the docker image from scratch passing the argument --no-cache to the docker build command?
Remember to change the ldap_username to your gaspar username.

**Comment by antonelepfl** [comment]
I'm getting an error in the building process
```
Step 63/77 : RUN spack install neurodamus@master~coreneuron
 ---> Running in 4bcccdde3905
==> Warning: No Spack mirrors are currently configured
==> Installing diffutils
==> Searching for binary cache of diffutils
==> No binary for diffutils found: installing from source
==> Fetching https://ftp.gnu.org/gnu/diffutils/diffutils-3.6.tar.xz
==> Staging archive: /home/mooc/softwares/spack/var/spack/stage/diffutils-3.6-cj7oc5hii5l5karvq4t5adcde2wueh6u/diffutils-3.6.tar.xz
==> Created stage in /home/mooc/softwares/spack/var/spack/stage/diffutils-3.6-cj7oc5hii5l5karvq4t5adcde2wueh6u
==> No patches needed for diffutils
==> Building diffutils [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed diffutils
  Fetch: 4.76s.  Build: 1m 1.10s.  Total: 1m 5.87s.
[+] /home/mooc/softwares/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/diffutils-3.6-cj7oc5
==> Warning: No Spack mirrors are currently configured
==> Installing bzip2
==> Searching for binary cache of bzip2
==> No binary for bzip2 found: installing from source
==> Fetching https://fossies.org/linux/misc/bzip2-1.0.6.tar.gz
==> Staging archive: /home/mooc/softwares/spack/var/spack/stage/bzip2-1.0.6-4orv5thyrbaev7odgm5g6r3ungzj5yxv/bzip2-1.0.6.tar.gz
==> Created stage in /home/mooc/softwares/spack/var/spack/stage/bzip2-1.0.6-4orv5thyrbaev7odgm5g6r3ungzj5yxv
==> Ran patch() for bzip2
==> Building bzip2 [Package]
==> Executing phase: 'install'
==> Successfully installed bzip2
  Fetch: 0.95s.  Build: 4.24s.  Total: 5.19s.
[+] /home/mooc/softwares/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/bzip2-1.0.6-4orv5t
==> Warning: No Spack mirrors are currently configured
==> Installing zlib
==> Searching for binary cache of zlib
==> No binary for zlib found: installing from source
curl: (18) transfer closed with 423253 bytes remaining to read
==> Fetching http://zlib.net/fossils/zlib-1.2.11.tar.gz
==> Fetching from http://zlib.net/fossils/zlib-1.2.11.tar.gz failed.
==> Error: FetchError: All fetchers failed for zlib-1.2.11-av3mmgf66zwic2i34n7o2oxtp4quy6pi

/home/mooc/softwares/spack/lib/spack/spack/package.py:925, in do_fetch:
        922                raise FetchError("Will not fetch %s" %
        923                                 self.spec.format('$_$@'), ck_msg)
        924
  >>    925        self.stage.create()
        926        self.stage.fetch(mirror_only)
        927        self._fetch_time = time.time() - start_time
        928


The command '/bin/sh -c spack install neurodamus@master~coreneuron' returned a non-zero code: 1
```

**Comment by pramodk** [comment]
```
curl: (18) transfer closed with 423253 bytes remaining to read
==> Fetching http://zlib.net/fossils/zlib-1.2.11.tar.gz
```
not sure why your docker not able to fetch tarballs....connectivity issue? could you try again

**Comment by antonelepfl** [comment]
Unfortunately this time the build was successful but I get the same error as before running the test simulation:
```
> docker run -i -t hipp-parallel /bin/bash -c 'cd $HOME/sim/build/circuitBuilding_1000neurons &&  . $SPACK_ROOT/share/spack/setup-env.sh && module av && module load neurodamus/master && mpiexec -n 6 --allow-run-as-root special $HOC_LIBRARY_PATH/init.hoc -mpi'

---------------------- /home/mooc/softwares/spack/share/spack/modules/linux-ubuntu18.04-x86_64 -----------------------
neurodamus/master          neurodamus-neocortex/0.1   neuron/2018-10
neurodamus-hippocampus/0.1 neurodamus-thalamus/0.1

[e844435497d7:00001] *** Process received signal ***
[e844435497d7:00001] Signal: Floating point exception (8)
[e844435497d7:00001] Signal code: Integer divide-by-zero (1)
[e844435497d7:00001] Failing at address: 0x7fa302dc20a0
[e844435497d7:00001] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7fa302b33890]
[e844435497d7:00001] [ 1] /usr/lib/x86_64-linux-gnu/libopen-pal.so.20(+0x820a0)[0x7fa302dc20a0]
[e844435497d7:00001] [ 2] /usr/lib/x86_64-linux-gnu/libopen-pal.so.20(opal_progress_set_event_poll_rate+0x17)[0x7fa302d66b47]
[e844435497d7:00001] [ 3] /usr/lib/x86_64-linux-gnu/libopen-pal.so.20(opal_progress_init+0x19)[0x7fa302d66b89]
[e844435497d7:00001] [ 4] /usr/lib/x86_64-linux-gnu/libopen-pal.so.20(opal_init+0x151)[0x7fa302d67c31]
[e844435497d7:00001] [ 5] mpiexec(+0x639a)[0x55be3745139a]
[e844435497d7:00001] [ 6] mpiexec(+0x5aea)[0x55be37450aea]
[e844435497d7:00001] [ 7] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7fa302751b97]
[e844435497d7:00001] [ 8] mpiexec(+0x59ea)[0x55be374509ea]
[e844435497d7:00001] *** End of error message ***
```

I have also tried to run my simulations with **neurodamus-hippocampus** and I get a similar error

**Comment by pramodk** [comment]
@antonelepfl : could you pass by when you have time?

**Comment by jorblancoa** [comment]
I just did the rebase from the openmpi branch and tested in my workstation.
Could you try again and let us know @antonelepfl?

**Comment by pramodk** [comment]
@jorblancoa : if this has been tested, we can merge this.

@antonelepfl : could you comment if this is working for you?

**Comment by antonelepfl** [comment]
Hi, I've been testing the performance and even though it seems like it writes in parallel at the end when I select with reports the cpus don't work at 100% during the simulation process making the simulation in general work slower than before.

Test with:
- Hippocampus Microcircuit
- 500 cells
- 300 ms
- 2 VM (8 CPUs per VM)

Docker Img | With Report | No Report
--| -- | --
Old (serial) | 33 min [log](https://github.com/BlueBrain/neuron-docker-spack/files/3007643/Serial-with-report.txt) | 10 min [log](https://github.com/BlueBrain/neuron-docker-spack/files/3007642/Serial-no-report.txt)
New (parallel) | 1:10 hour [log](https://github.com/BlueBrain/neuron-docker-spack/files/3007645/Parallel-with-report.txt)| 15 min [log](https://github.com/BlueBrain/neuron-docker-spack/files/3007646/Parallel-no-report.txt)

As I mentioned during the simulation I see the that the CPUs usage goes down (sometimes to 0%).
I have attached the log of the different jobs and you will see that for example
```
Completed =91% ETA =0:04:36  
... After 3:30 min ...
Completed =92% ETA =0:04:32  
```
That means that at some point one of the VMs decreased the CPU usage.
This **does not** happen if I skip the reports.

Just for the record, if we run the simulation in **one VM** the results are the followings:

Docker Img | With Report | No Report
--| -- | --
Parallel | 49m:50s |   22m:30s
Serial | 25m:30s | 19m:34s

**Comment by pramodk** [comment]
@antonelepfl : for performance we can debug that separately.

**Comment by antonelepfl** [comment]
Ok agree. So then I guess we can merge this PR.

**Comment by pramodk** [comment]
@jorblancoa : anything pending here? if not, could you merge this?

**Pull request merged**